src/host/services/IMUService.cpp

```diff
@@ -23,41 +23,138 @@
+    delay(50);
+    bno055.setMode( bno055.OPERATION_MODE_NDOF ); // OPERATION_MODE_NDOF_FMC_OFF
+    delay(1000);
+    if (recallCalibration())
+      DEBUG("Success!");
   }
 }
 
 void IMUService::loop() {
-  bno055.getCalibration(&sysCalib, &gyroCalib, &accelCalib, &magCalib);
+  bno055.getCalibration(&_sysCalib, &_gyroCalib, &_accelCalib, &_magCalib);
 
   eulerAngles = bno055.getVector(Adafruit_BNO055::VECTOR_EULER);
 
-  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", sysCalib, accelCalib, gyroCalib, magCalib);
+  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", _sysCalib, _accelCalib, _gyroCalib, _magCalib);
   //DEBUG("Attitude roll: %f pitch: %f  Mag heading: %f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
 
   SKUpdateStatic<2> update;
   // Note: We could calculate yaw as the difference between the Magnetic
   // Heading and the Course Over Ground Magnetic.
 
-  /* if orientation == MOUNTED_ON_PORT_HULL */
-  double roll, pitch, heading;
-  roll = eulerAngles.z();
-  pitch = eulerAngles.y();
-  heading = fmod(eulerAngles.x() + 270, 360);
+  /* if orientation == MOUNTED_ON_STARBOARD_HULL */
+  _roll = eulerAngles.z() + _offsetRoll;
+  _pitch = eulerAngles.y() + _offsetPitch;
+  _heading = fmod(eulerAngles.x() + 270, 360);
+  //DEBUG("Attitude heel: %.3f pitch: %.3f  Mag heading: %.3f", _roll, _pitch, _heading);
+
+  if (_magCalib >= cfHdgMinCal) {
+    update.setNavigationHeadingMagnetic(SKDegToRad(_heading));
+    _skHub.publish(update);
+  }
 
-  if (sysCalib == 3) {
-    update.setNavigationAttitude(SKTypeAttitude(/* roll */ SKDegToRad(roll), /* pitch */ SKDegToRad(pitch), /* yaw */ 0));
-    update.setNavigationHeadingMagnetic(SKDegToRad(heading));
+	if (_accelCalib >= cfHeelPitchMinCal && _gyroCalib >= cfHeelPitchMinCal) {
+    update.setNavigationAttitude(SKTypeAttitude(/* roll */ SKDegToRad(_roll), /* pitch */ SKDegToRad(_pitch), /* yaw */ 0));
     _skHub.publish(update);
   }
 }
+
+void IMUService::getLastValues(int &accelCalibration, double &pitch, double &roll, int &magCalibration, double &heading){
+  accelCalibration = _accelCalib;
+  pitch = _pitch;
+  roll = _roll;
+  magCalibration = _magCalib;
+  heading = _heading;
+}
+
+//  Offset for Heel=0 and Pitch=0 with long button press in IMUMonitorPage
+void IMUService::setOffset() {
+  // FIXIT: Why it is called at startup of KBox? event longClick coming?
+  if ( millis() > 10000 ) {
+    _offsetRoll = _roll * (-1);
+    _offsetPitch = _pitch * (-1);
+    DEBUG("Offset changed: Heel -> %.3f | Pitch -> %.3f", _offsetRoll, _offsetPitch);
+    // TODO: angle of heel and pitch must be smaller than 90°, check overflow!
+  }
+}
+
+bool IMUService::saveCalibration() {
+
+   int _eeAddress = 0;
+   long bnoID;
+   sensor_t sensor;
+   adafruit_bno055_offsets_t newCalib;
+   bno055.getSensorOffsets(newCalib);
+
+   DEBUG("\n\nStoring calibration data to EEPROM...");
+
+   bno055.getSensor(&sensor);
+   bnoID = sensor.sensor_id;
+
+   EEPROM.put(_eeAddress, bnoID);
```

Comment:
Instead of writing directly to the EEPROM could you try to re-use https://github.com/sarfata/kbox-firmware/blob/master/src/host/util/PersistentStorage.cpp?

This way we have only one place where we manage what is stored in EEPROM and where. It will help us make sure that we are not overwriting our own data and if we want to choose to another type of storage (on another micro-controller that does not have EEPROM for example) then we can do it in one place only.

Reply:
OK, makes a lot of sense. Will try to do it.